Target
______


Parameters: Less than 10,000

Data Augmentations: No

Regularization: Yes

LR Scheduler: No

No. of Epochs: 14

In [1]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torchsummary import summary

In [2]:
train = datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor()]))

  0%|          | 0/9912422 [00:00<?, ?it/s]

Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw



  0%|          | 0/28881 [00:00<?, ?it/s]

Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw



  0%|          | 0/1648877 [00:00<?, ?it/s]

Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw



  0%|          | 0/4542 [00:00<?, ?it/s]

Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw



/usr/local/lib/python3.7/dist-packages/torchvision/datasets/mnist.py:498: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:180.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


In [3]:
torch.manual_seed(1)
batch_size = 64

use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")

kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}

train_data = datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))]))
train_loader = torch.utils.data.DataLoader(train_data,batch_size=batch_size,
                                           shuffle=True, **kwargs)

test_data = datasets.MNIST('../data', train=False, transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))]))
test_loader = torch.utils.data.DataLoader(test_data, batch_size=batch_size,
                                          shuffle=True, **kwargs)

In [4]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()

        self.layer1 = nn.Sequential(
            # 28x28x1 -> 26x26x4
            nn.Conv2d(in_channels=1, out_channels=4, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(num_features=4),
            nn.ReLU(),
            nn.Dropout(0.043),
            # 26x26x4 -> 24x24x8
            nn.Conv2d(in_channels=4, out_channels=8, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(num_features=8),
            nn.ReLU(),
            nn.Dropout(0.043),
            # 24x24x8 -> 12x12x8
            nn.MaxPool2d(kernel_size=2, stride=2)
        )


        self.layer2 =  nn.Sequential(
            # 12x12x8 -> 10x10x16
            nn.Conv2d(in_channels=8, out_channels=16, kernel_size=3, bias=False),
            nn.BatchNorm2d(num_features=16),
            nn.ReLU(),
            nn.Dropout(0.043),
            # 10x10x16 -> 8x8x20
            nn.Conv2d(in_channels=16, out_channels=20, kernel_size=3, bias=False),
            nn.BatchNorm2d(num_features=20),
            nn.ReLU(),
            nn.Dropout(0.043),
            # 8x8x20 -> 6x6x24
            nn.Conv2d(in_channels=20, out_channels=24, kernel_size=3, bias=False),
            nn.BatchNorm2d(num_features=24),
            nn.ReLU(),
            nn.Dropout(0.043),
            # 6x6x24 -> 6x6x10
            nn.Conv2d(in_channels=24, out_channels=10, kernel_size=1, bias=False)
        )

        self.gap = nn.AdaptiveAvgPool2d(1)

    def forward(self, x):
        x = self.layer1(x)
        x = self.layer2(x)
        x = self.gap(x)
        x = x.view(-1,10)

        return F.log_softmax(x, dim=1)

In [5]:
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 4, 28, 28]              36
       BatchNorm2d-2            [-1, 4, 28, 28]               8
              ReLU-3            [-1, 4, 28, 28]               0
           Dropout-4            [-1, 4, 28, 28]               0
            Conv2d-5            [-1, 8, 28, 28]             288
       BatchNorm2d-6            [-1, 8, 28, 28]              16
              ReLU-7            [-1, 8, 28, 28]               0
           Dropout-8            [-1, 8, 28, 28]               0
         MaxPool2d-9            [-1, 8, 14, 14]               0
           Conv2d-10           [-1, 16, 12, 12]           1,152
      BatchNorm2d-11           [-1, 16, 12, 12]              32
             ReLU-12           [-1, 16, 12, 12]               0
          Dropout-13           [-1, 16, 12, 12]               0
           Conv2d-14           [-1, 20,

/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)


In [6]:
from tqdm import tqdm

train_losses = []
test_losses = []
train_acc = []
test_acc = []

def train(model, device, train_loader, optimizer, epoch):
  model.train()
  pbar = tqdm(train_loader)
  correct = 0
  processed = 0
  for batch_idx, (data, target) in enumerate(pbar):
    # get samples
    data, target = data.to(device), target.to(device)

    # Init
    optimizer.zero_grad()
    # In PyTorch, we need to set the gradients to zero before starting to do backpropragation because PyTorch accumulates the gradients on subsequent backward passes. 
    # Because of this, when you start your training loop, ideally you should zero out the gradients so that you do the parameter update correctly.

    # Predict
    y_pred = model(data)

    # Calculate loss
    loss = F.nll_loss(y_pred, target)
    train_losses.append(loss)

    # Backpropagation
    loss.backward()
    optimizer.step()

    # Update pbar-tqdm
    
    pred = y_pred.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
    correct += pred.eq(target.view_as(pred)).sum().item()
    processed += len(data)

    pbar.set_description(desc= f'Loss={loss.item()} Batch_id={batch_idx} Accuracy={100*correct/processed:0.2f}')
    train_acc.append(100*correct/processed)

def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)
    test_losses.append(test_loss)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))
    
    test_acc.append(100. * correct / len(test_loader.dataset))

In [7]:
model =  Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
EPOCHS = 14
for epoch in range(EPOCHS):
    print("EPOCH:", epoch)
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

EPOCH: 0


Loss=0.12439046055078506 Batch_id=937 Accuracy=83.90: 100%|██████████| 938/938 [01:00<00:00, 15.46it/s]



Test set: Average loss: 0.1352, Accuracy: 9645/10000 (96.45%)

EPOCH: 1


Loss=0.03918438032269478 Batch_id=937 Accuracy=97.19: 100%|██████████| 938/938 [01:00<00:00, 15.42it/s]



Test set: Average loss: 0.0648, Accuracy: 9818/10000 (98.18%)

EPOCH: 2


Loss=0.10471709817647934 Batch_id=937 Accuracy=97.96: 100%|██████████| 938/938 [01:00<00:00, 15.42it/s]



Test set: Average loss: 0.0439, Accuracy: 9872/10000 (98.72%)

EPOCH: 3


Loss=0.02116960659623146 Batch_id=937 Accuracy=98.33: 100%|██████████| 938/938 [01:00<00:00, 15.57it/s]



Test set: Average loss: 0.0413, Accuracy: 9873/10000 (98.73%)

EPOCH: 4


Loss=0.09054634720087051 Batch_id=937 Accuracy=98.42: 100%|██████████| 938/938 [01:00<00:00, 15.52it/s]



Test set: Average loss: 0.0442, Accuracy: 9871/10000 (98.71%)

EPOCH: 5


Loss=0.13853786885738373 Batch_id=937 Accuracy=98.51: 100%|██████████| 938/938 [01:00<00:00, 15.55it/s]



Test set: Average loss: 0.0387, Accuracy: 9881/10000 (98.81%)

EPOCH: 6


Loss=0.19236484169960022 Batch_id=937 Accuracy=98.73: 100%|██████████| 938/938 [01:01<00:00, 15.32it/s]



Test set: Average loss: 0.0322, Accuracy: 9903/10000 (99.03%)

EPOCH: 7


Loss=0.04331258684396744 Batch_id=937 Accuracy=98.78: 100%|██████████| 938/938 [01:00<00:00, 15.40it/s]



Test set: Average loss: 0.0319, Accuracy: 9893/10000 (98.93%)

EPOCH: 8


Loss=0.00943798664957285 Batch_id=937 Accuracy=98.85: 100%|██████████| 938/938 [01:00<00:00, 15.43it/s]



Test set: Average loss: 0.0323, Accuracy: 9911/10000 (99.11%)

EPOCH: 9


Loss=0.025353562086820602 Batch_id=937 Accuracy=98.92: 100%|██████████| 938/938 [01:00<00:00, 15.41it/s]



Test set: Average loss: 0.0345, Accuracy: 9889/10000 (98.89%)

EPOCH: 10


Loss=0.058327995240688324 Batch_id=937 Accuracy=98.95: 100%|██████████| 938/938 [01:01<00:00, 15.16it/s]



Test set: Average loss: 0.0263, Accuracy: 9904/10000 (99.04%)

EPOCH: 11


Loss=0.034403931349515915 Batch_id=937 Accuracy=98.97: 100%|██████████| 938/938 [01:03<00:00, 14.71it/s]



Test set: Average loss: 0.0268, Accuracy: 9903/10000 (99.03%)

EPOCH: 12


Loss=0.16951978206634521 Batch_id=937 Accuracy=99.05: 100%|██████████| 938/938 [01:05<00:00, 14.40it/s]



Test set: Average loss: 0.0264, Accuracy: 9910/10000 (99.10%)

EPOCH: 13


Loss=0.0048132590018212795 Batch_id=937 Accuracy=99.09: 100%|██████████| 938/938 [01:05<00:00, 14.36it/s]



Test set: Average loss: 0.0284, Accuracy: 9909/10000 (99.09%)



Total Parameters used: 9,060

Best Train Accuracy: 99.05%

Best Test Accuracy: 99.10%

Consistency: Did not achieve any consistency with test accuracy greater than 99.10
Added Dropout Layer to the network and remove FC layers .

we can see some underfitting in the model in the next model lets try to use Data Augmentation and Step wise LR .
